large_image/tilesource/jupyter.py

```diff
@@ -393,19 +398,59 @@ def from_map(self, coordinate: Union[List[float], Tuple[float, float]]) -> Tuple
             return transf.transform(x, y)
         return x, self._metadata['sizeY'] - y
 
-    def update_frame(self, event, **kwargs):
-        frame = int(event.get('new'))
+    def update_frame(self, frame, style, **kwargs):
         if self._layer:
-            if 'frame=' in self._layer.url:
-                self._layer.url = re.sub(r'frame=(\d+)', f'frame={frame}', self._layer.url)
-            else:
-                if '?' in self._layer.url:
-                    self._layer.url = self._layer.url.replace('?', f'?frame={frame}&')
-                else:
-                    self._layer.url += f'?frame={frame}'
-
+            parsed_url = urlparse(self._layer.url)
+            query = parsed_url.query
+            query = {k: v[0] for k, v in parse_qs(query).items()}
+            query.update(dict(
+                frame=frame,
+                style=json.dumps(style),
+            ))
+            query_string = urlencode(query, quote_via=quote, safe='{}')
+            self._layer.url = urlunparse((
+                parsed_url.scheme, parsed_url.netloc,
+                parsed_url.path, parsed_url.params,
+                query_string, parsed_url.fragment,
+            ))
             self._layer.redraw()
 
+    def get_frame_histogram(self, query):
+        if self._layer is not None:
+            frame = query.get('frame')
+            frame_histograms = self.frame_selector.frameHistograms or {}
```

Comment:
Why do you copy this and modify the copy?  If there are concurrent histogram requests, won't this end up just keeping the last one?

Reply:
The copy is done so that we set `self.frame_selector.frame_histograms` to a new object to trigger the Vue watcher that detects changes to that object. Without the copy statement, only the first histogram will be shown and Vue cannot detect subsequent updates. I haven't encountered a case where concurrent requests overwrite one another, but maybe some extra bookkeeping on this class could resolve that.

Reply:
Extra bookkeeping added in [303ebe3](https://github.com/girder/large_image/pull/1738/commits/303ebe30d932c8c0c69fa020ad811c12cd719b72).